pkg/controllermanager/controller/seed/seed_reconcile.go

```diff
@@ -0,0 +1,204 @@
+	clientMap    clientmap.ClientMap
+	secretLister corev1listers.SecretLister
+	seedLister   gardencorelisters.SeedLister
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	seedObj, err := r.seedLister.Get(req.Name)
+	if apierrors.IsNotFound(err) {
+		logger.Logger.Infof("[SEED] Stopping operations for Seed %s since it has been deleted", req.Name)
+		return reconcileResult(nil)
+	}
+	if err != nil {
+		logger.Logger.Infof("[SEED] %s - unable to retrieve object from store: %v", req.Name, err)
+		return reconcileResult(err)
+	}
+
+	gardenClient, err := r.clientMap.GetClient(ctx, keys.ForGarden())
+	if err != nil {
+		return reconcileResult(fmt.Errorf("failed to get garden client: %w", err))
+	}
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("seed-%s", seedObj.Name),
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(seedObj, gardencorev1beta1.SchemeGroupVersion.WithKind("Seed")),
+			},
+		},
+	}
+
+	// TODO: use controllerutil.CreateOrUpdate once cache is graduated to beta https://github.com/gardener/gardener/issues/2822
+	if err := gardenClient.Client().Create(ctx, namespace); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return reconcileResult(err)
+		}
+		if _, err2 := kutil.TryUpdateNamespace(ctx, gardenClient.Kubernetes(), retry.DefaultRetry, metav1.ObjectMeta{Name: namespace.Name}, func(ns *corev1.Namespace) (*corev1.Namespace, error) {
+			ns.OwnerReferences = namespace.OwnerReferences
+			return ns, nil
+		}); err2 != nil {
+			return reconcileResult(err2)
+		}
+	}
+
+	syncedSecrets, err := r.syncGardenSecrets(ctx, gardenClient, namespace)
+	if err != nil {
+		return reconcileResult(fmt.Errorf("failed to sync garden secrets: %v", err))
+	}
+
+	if err := r.cleanupStaleSecrets(ctx, gardenClient, syncedSecrets, namespace.Name); err != nil {
+		return reconcileResult(fmt.Errorf("failed to clean up secrets in seed namespace: %v", err))
+	}
+
+	return reconcileResult(nil)
+}
+
+func (r *reconciler) cleanupStaleSecrets(ctx context.Context, gardenClient kubernetes.Interface, existingSecrets []string, namespace string) error {
+	var fns []flow.TaskFn
+	exclude := sets.NewString(existingSecrets...)
+
+	secrets, err := r.secretLister.Secrets(namespace).List(gardenRoleSelector)
+	if err != nil {
+		return err
+	}
+
+	for _, s := range secrets {
+		secret := s
+		if exclude.Has(secret.Name) {
+			continue
+		}
+		fns = append(fns, func(ctx context.Context) error {
+			if err := gardenClient.Client().Delete(ctx, secret); client.IgnoreNotFound(err) != nil {
+				return err
+			}
+			return nil
+		})
+	}
+
+	return flow.Parallel(fns...)(ctx)
+}
+
+func (r *reconciler) syncGardenSecrets(ctx context.Context, gardenClient kubernetes.Interface, namespace *corev1.Namespace) ([]string, error) {
+	secretsGardenRole, err := r.secretLister.Secrets(v1beta1constants.GardenNamespace).List(gardenRoleSelector)
```

Comment:
> Just to verify my understanding: we are preventing to sync the shoot secrets (kubeconfig, ssh-keypair and so) of shoots in the garden namespace

Actually, we do this for all namespaces.

So, if we wouldn't do it there, then the secretLister here would also return the shoot secrets, correct?

Yes, that's correct. Do you think that's too implicit?

Reply:
> Do you think that's too implicit?

Hmm, maybe it would be better to have an explicit selector here.
I'm just a bit afraid, that we might miss the implications of changing the special informer creation and end up with sensitive secrets synced across all seed namespaces...

WDYT about using a label selector for the domain secrets here explicitly? It shouldn't hurt, right?

Reply:
There are also other secrets that need to be synced (alerting smtp, openvpn dh key, ...)

Reply:
Ok, then what about using the same selector like the special informer (i.e. unselecting all shoot secrets)?

Reply:
Okay, good idea. I've added it to the PR.